docs/lcdvc.1.in
.\" This man page is released under the GNU General Public License.
.\" Refer to the COPYING file distributed with this package.
.TH "lcdvc" "1" "12 June 2009" "LCDproc" "LCDproc suite"
.SH "NAME"
.LP
lcdvc \- LCDproc virtual console
.SH "SYNOPSIS"
.LP
lcdvc
[\fB\-fh\fR]
[\fB\-a\fR \fIaddr\fR]
[\fB\-c\fR \fIfile\fR]
[\fB\-p\fR \fIport\fR]
[\fB\-r\fR \fIlevel\fR]
[\fB\-s\fR \fIbool\fR]
.SH "DESCRIPTION"
.LP
This program shows content from a system console on a LCD controlled by
\fBLCDd\fR. As the console is usually larger than a LCD, the area shown on
the LCD screen can be scrolled using keys attached to the display.
.LP
Additionally it has an autoscroll feature, which is on by default and can
only be disabled by using \fBlcdvc\fR's menu.
.LP
By default it uses \fI/dev/vcs\fR and \fI/dev/vcsa\fR devices as the
console to show. Other console devices can be set in the configuration file
\fI@SYSCONFDIR@/lcdvc.conf\fR. The keys used for scrolling can be set in the config
file as well.
.SH "OPTIONS"
.LP
The following options can be set in the configuration file
\fI@SYSCONFDIR@/lcdvc.conf\fR as well, but those given on the command line take
precedence:
.TP
\fB\-a\fR \fIaddr\fR
DNS name or IP address of the LCDd server (default localhost).
.TP
\fB\-c\fR \fIfile\fR
Specify configuration file.
.TP
\fB\-f\fR
Run in foreground.
.TP
\fB\-h\fR
Output help and exit.
.TP
\fB\-p\fR \fIport\fR
Port of the LCDd server (default 13666).
.TP
\fB\-r\fR \fIlevel\fR
Set reporting level (0\-5, default: 2)
.TP
\fB\-s\fR \fIbool\fR
Report to syslog (1) or stderr (0, default).
.SH "FILES"
.LP
\fI@SYSCONFDIR@/lcdvc.conf\fP
.br
\fI/dev/vcs\fP
.br
\fI/dev/vcsa\fP
.SH "AUTHORS"
.LP
\fBlcdvc\fR was originally written by Joris Robijn, modified by Peter Marschall.
.LP
This man page was written by Markus Dolze.
.SH "SEE ALSO"
.LP
LCDd(8) vcs(4)
.SH "LICENSE"
\fBlcdvc\fR is released under the GNU General Public License, version 2.


ChangeLog
Recent changes:

Key:
     - Something removed
     + Something added
     * Something changed / fixed

v.0.5dev (ongoing development)
 + sed1330 driver: Add support for HG25504 (L. Lagendijk)
 * Build system: Fix xosd detection
 * Reset ccmode on screen clear for NoritakeVFD, tyan driver
 * All drivers: Refactor CGmode enum into lcd.h
 * hd44780: New (possible faster) screen update algorithm
 + Add an option to start LCDd with screen rotation disabled (M. Zanetti)
 * Rewrite of the t6963 driver fixing several problems (see commit note).
 * picolcd: Add OffBrightness (S. Crane), more icons and fix bignum
 * lcdproc: Fix a single dot appearing on line 5 of the D screen
 * lcdproc: Make the iface screen more readable on 2x16 displays
 * All drivers: sync functions with prototypes
 * Port remaining drivers to 0.5 API (glk, lcdm001, stv5730)
 * Pyramid: Only send LED state changes, doxygen-ize
 * sed1520: Make it work with 68-family style interface, convert to 0.5 API
 * hd44780/imon: Add charmap for NEC uPD16314 and make it configurable
 * Pyramid: Convert set_char to 8 byte bitmap, add support for bignum
 * Convert set_char of drivers bayrad, lb216, lcterm, mtc_s16209x, wirz-sli
 * CFontzPacket: Fix not using select() even if its available
 * CFontzPacket: Use cellwidth = 5 on CFA-631 and CFA-635 too
 * t6963: Convert font to 8 byte bitmap
 * convert i2500vfd and mdm166a driver to use 8 byte bitmap custom character
 * ula200: convert custom characters to 8 byte bitmap
 * Build system: Fix trying to install svga driver twice
 + imon: Add bignum support (Y. Scheglyuk)
 * hd44780/ethlcd: fix LCDd hang up on network error (exits now) (M. Bialonczyk)

v0.5.4
 * Update driver includes and LDFLAGS (fixed glk and bayrad binding error)
 * Modify all drivers to use server/drivers/report.h instead of shared/report.h
 * Make all drivers using parallel port fail if they cannot get IO permission
 * Include <errno.h> instead of <sys/errno.h> everywhere (found by S. Klauer)
 * configure: Fix de-selecting of drivers
 * Build system: Add a check testing if SA_RESTART flag is available.
 * picolcd driver: Fix RC-5 for picoLCD 20x2 (A. van Schie)
 * Introduce shared/defines.h to collect commonly used macros
 + new driver: SureElec for devices made by SURE Electronics (Laurent Latil)
 * hd44780-4bit: Fix bug not checking for keypad rows 7-10 correctly
 * Modify serial drivers to use cfmakeraw
 + hd44780: Add a 'none' charmap which does not replace any character
 * hd44780: Change mapping for spanish 'n with tilde' characters
 * hd44780: Exclude pin for switchable backlight from keypad scanning
 * server core: New network input buffering
 * hd44780: Extended subdriver API and performance improvement for lcd2usb
 * hd44780: Add KOI8-R character mapping (Yura Scheglyuk)
 * shuttleVFD: Add support for devices with USB VID 0x1308 again
 * CFontzPacket: Add CFA-533. Use HD44780 character mapping for CFA533/633.
 * CFontzPacket: Report cell size correctly, use block character for icon.
 * CFontz633: Deprecate this driver. Use CFontzPacket with Model=633 instead!
 - Remove native win32 support.
 * picolcd: Use module output function to control key LEDs (Martin Jones)
 + hd44780: Added 'usbtiny' connection type (Siarhei Herasimchuk)
 * SureElec: Fix stall on startup
 + new driver: MDM166A for Futaba/Targa USB VFD (Christoph Rasim)
 * picolcd: Fix backlight and contrast handling (M. T. Jones)
 * picolcd: Fix heartbeat icon messing up with vbar
 * imonlcd: Fix possible lockup in output (icons) function (E. Pooch)
 * pyramid: Fix buffer overflow in set_leds, incorrect escaping, and icons

v0.5.3
 + lcdexec: notification when called program finishes
 + xosd driver: offset from top-left corner
 + LCDd: Hello= config option for startup message
 * glk driver: bignum support fixed
 * update local snprintf to the last one, which is GPL
 - server core: cleanup - remove unnecessary cruft
 * server core: refactor rendering code
 + LCDd: config options TitleSpeed= and Heartbeat=
 * overhauled CwLnx driver (Dave Platt)
 * overhauled picolcd driver (Nicu Pavel)
 + add ServerScreen=blank option value
 * MtxOrb driver: fix VFD/VKD support (Ethan Dicks)
 + new driver lis for VLSystem L.I.S VFD (Daryl Fonseca-Holt)
 * serialPOS driver: input support, cleanup (Eric Pooch)
 + new driver shuttleVFD for USB-based Shuttle VFDs (Thien Vu)
 * enable building lcdproc client on FreeBSD AMD64 platform (M. Dolze)
 * Autoconf fixes for Net/FreeBSD (M. Dolze)
 * fixes for the lcpdroc client iface screen on *BSD (M.Dolze)
 + key support for hd44780/lcd2usb (M. Dolze)
 + option Priority for NoritakeVFD (idea: Richard Muratti)
 * overhauled NoritakeVFD driver (idea: Richard Muratti)
 * allow wrapping slider menu items with <= 4 keys (M. Dolze)
 * a bit more Doxygen-ation and documentation
 * hd44780 driver: make ConnectionType 4bit more fool-proof (M. Dolze)
 * picolcd driver: overhaul USB init for more portability (M. Dolze)
 + curses driver: new option DrawBorder (Bruno Schwander)
 * avoid sending duplicate "success" messages in response to menu_add_item
 + lcdexec: parameter support using widgets
 * server core: display values in menu for all input MenuEntry types
 * hd44780 driver: reorganize ConnectionType recognition
 + hd44780 driver: new ConnectionType mplay
 + hd44780 driver: get/set contrast & brightness support for lcd2usb (M. Dolze)
 + hd44780 driver: reporting in ConnectionType functions (Thomas Jarosch)
 * clean up build system (T. Jarosch, with tests/fixes by M. Dolze)
 + hd44780 driver: new ConnectionType ftdi (Thomas Jarosch)
 * shuttleVFD driver: fix VendorID, support newer models (Miska Sulander)
 + i2500vfd driver for a graphical Noritake VFD (Thomas Jarosch)
 * hd44780 driver: more complete initialization (idea by Pierre Ossman)
 * hd44780 driver: "euro" character mapping table (Markus Dolze)
 * CwLnx driver: support for CW12832
 + new driver mx5000 for Logitech MX5000 keyboards (Christian Jodar)

clients/lcdvc/lcdvc.c
/** \file clients/lcdvc/lcdvc.c
 * Main file for \c lcdvc, the LCDproc virtual console.
 */

/*-
 * This file is part of lcdvc, an LCDproc client.
 *
 * This file is released under the GNU General Public License. Refer to the
 * COPYING file distributed with this package.
 *
 * Copyright (c) 2002, Joris Robijn
 *               2006-2008, Peter Marschall
 */

#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <signal.h>
#include <fcntl.h>
#include <errno.h>
#include <stdlib.h>

#include "getopt.h"

#include "shared/report.h"
#include "shared/configfile.h"
#include "shared/sockets.h"
#include "lcd_link.h"
#include "vc_link.h"
#include "lcdvc.h"

#if !defined(SYSCONFDIR)
# define SYSCONFDIR	"/etc"
#endif
#if !defined(PIDFILEDIR)
# define PIDFILEDIR	"/var/run"
#endif

#define DEFAULT_CONFIGFILE	SYSCONFDIR "/lcdvc.conf"
#define DEFAULT_PIDFILE		PIDFILEDIR "/lcdvc.pid"


char *help_text =
"lcdvc - LCDproc virtual console\n"
"\n"
"Copyright (c) 2002, Joris Robijn, 2006-2008 Peter Marschall.\n"
"This program is released under the terms of the GNU General Public License.\n"
"\n"
"Usage: lcdvc [<options>]\n"
"  where <options> are:\n"
"    -c <file>           Specify configuration file ["DEFAULT_CONFIGFILE"]\n"
"    -a <address>        DNS name or IP address of the LCDd server [localhost]\n"
"    -p <port>           port of the LCDd server [13666]\n"
"    -f                  Run in foreground\n"
"    -r <level>          Set reporting level (0-5) [2: errors and warnings]\n"
"    -s <0|1>            Report to syslog (1) or stderr (0, default)\n"
"    -h                  Show this help\n";

/* Variables set by config */
int foreground = FALSE;
static int report_level = UNSET_INT;
static int report_dest = UNSET_INT;
char *vcsa_device = UNSET_STR;
char *vcs_device = UNSET_STR;
char *keys[4];

/* Other global variables */
char *progname = "lcdvc";
char *configfile = UNSET_STR;
char *pidfile = NULL;
int pidfile_written = FALSE;

int Quit = 0;			/**< indicate end of main loop */

/* Function prototypes */
static void exit_program(int val);
static int process_command_line(int argc, char **argv);
static int process_configfile(char *configfile);
static int main_loop(void);


int main(int argc, char **argv)
{
	int e = 0;
	struct sigaction sa;

	CHAIN( e, process_command_line( argc, argv ));
	if (strcmp( configfile, UNSET_STR ) == 0) {
		configfile = DEFAULT_CONFIGFILE;
	}
	CHAIN( e, process_configfile( configfile ));
	if (report_dest == UNSET_INT || report_level == UNSET_INT) {
		report_dest = RPT_DEST_STDERR;
		report_level = RPT_ERR;
	}
	set_reporting( progname, report_level, report_dest );

	CHAIN( e, open_vcs() );
	CHAIN( e, setup_connection() );
	CHAIN_END( e );

	if (foreground != TRUE) {
		if (daemon(1,1) != 0) {
			report(RPT_ERR, "Error: daemonize failed");
		}

		if (pidfile != NULL) {
			FILE *pidf = fopen(pidfile, "w");

			if (pidf) {
				fprintf(pidf, "%d\n", (int) getpid());
				fclose(pidf);
				pidfile_written = TRUE;
			} else {
				fprintf(stderr, "Error creating pidfile %s: %s\n",
					pidfile, strerror(errno));
				return(EXIT_FAILURE);
			}
		}
	}

	/* setup signal handlers for common signals */
	sigemptyset(&sa.sa_mask);
#ifdef HAVE_SA_RESTART
	sa.sa_flags = SA_RESTART;
#endif
	sa.sa_handler = exit_program;
	sigaction(SIGINT, &sa, NULL);	// Ctrl-C
	sigaction(SIGTERM, &sa, NULL);	// "regular" kill
	sigaction(SIGHUP, &sa, NULL);	// kill -HUP
	sigaction(SIGPIPE, &sa, NULL);	// write to closed socket
	sigaction(SIGKILL, &sa, NULL);	// kill -9 [cannot be trapped; but ...]

	main_loop();

	exit_program(EXIT_SUCCESS);

	/* NOTREACHED */
	return EXIT_SUCCESS;
}


static void exit_program(int val)
{
	//printf("exit program\n");
	Quit = 1;
	teardown_connection();
	if ((foreground != TRUE) && (pidfile != NULL) && (pidfile_written == TRUE))
		unlink(pidfile);
	exit(val);
}


static int process_command_line(int argc, char **argv)
{
	int c;
	int error = 0;

	/* No error output from getopt */
	opterr = 0;

	while ((c = getopt(argc, argv, "hc:a:p:fr:s:")) > 0) {
		char *end;
		int temp_int;

		switch (c) {
		  case 'h':
			fprintf(stderr, "%s", help_text);
			exit(EXIT_SUCCESS);
			/* NOTREACHED */
		  case 'c':
			configfile = strdup(optarg);
			break;
		  case 'a':
			address = strdup(optarg);
			break;
		  case 'p':
			temp_int = strtol(optarg, &end, 0);
			if ((*optarg != '\0') && (*end == '\0') &&
			     (temp_int > 0) && (temp_int <= 0xFFFF)) {
				port = temp_int;
			} else {
				report(RPT_ERR, "Illegal port value %s", optarg);
				error = -1;
			}
			break;
		  case 'f':
			foreground = TRUE;
			break;
		  case 'r':
			temp_int = strtol(optarg, &end, 0);
			if ((*optarg != '\0') && (*end == '\0') && (temp_int >= 0)) {
				report_level = temp_int;
			} else {
				report(RPT_ERR, "Illegal report level value %s", optarg);
				error = -1;
			}
			break;
		  case 's':
			temp_int = strtol(optarg, &end, 0);
			if ((*optarg != '\0') && (*end == '\0') && (temp_int >= 0)) {
				report_dest = (temp_int ? RPT_DEST_SYSLOG : RPT_DEST_STDERR);
			} else {
				report(RPT_ERR, "Illegal log destination value %s", optarg);
				error = -1;
			}
			break;
		  case ':':
			report(RPT_ERR, "Missing option argument for %c", optopt);
			error = -1;
			break;
		  case '?':
		  default:
			report(RPT_ERR, "Unknown option: %c", optopt);
			error = -1;
			break;
        	}
        }
	return error;
}


static int process_configfile(char *configfile)
{
	if (strcmp(configfile, UNSET_STR) == 0) {
		configfile = DEFAULT_CONFIGFILE;
	}
	if (config_read_file( configfile ) < 0) {
		report( RPT_WARNING, "Could not read config file: %s", configfile);
	}

	if (strcmp(address, UNSET_STR ) == 0) {
		address = strdup(config_get_string( progname, "Address", 0, "localhost"));
	}
	if (port == UNSET_INT) {
		port = config_get_int(progname, "Port", 0, 13666);
	}
	if (report_level == UNSET_INT ) {
		report_level = config_get_int(progname, "ReportLevel", 0, RPT_WARNING);
	}
	if (report_dest == UNSET_INT) {
		if (config_get_bool(progname, "ReportToSyslog", 0, 0)) {
			report_dest = RPT_DEST_SYSLOG;
		} else {
			report_dest = RPT_DEST_STDERR;
		}
	}
	if (foreground != TRUE) {
		foreground = config_get_bool(progname, "Foreground", 0, FALSE);
	}
	if (pidfile == NULL) {
		pidfile = strdup(config_get_string(progname, "PidFile", 0, DEFAULT_PIDFILE));
	}

	vcs_device = strdup(config_get_string(progname, "vcsDevice", 0, "/dev/vcs"));
	vcsa_device = strdup(config_get_string(progname, "vcsaDevice", 0, "/dev/vcsa"));

	keys[0] = strdup(config_get_string( progname, "UpKey", 0, "Up"));
	keys[1] = strdup(config_get_string( progname, "DownKey", 0, "Down"));
	keys[2] = strdup(config_get_string( progname, "LeftKey", 0, "Left"));
	keys[3] = strdup(config_get_string( progname, "RightKey", 0, "Right"));

	return 0;
}


static int main_loop(void)
{
	int num_bytes;
	char buf[80];
	short w = 0;

	/* Continuously check if we get a menu event... */

	while (!Quit && ((num_bytes = read_response(buf, sizeof(buf)-1)) >= 0)) {
		if (num_bytes != 0) {
			process_response(buf);
		}
		else {
			read_vcdata();
			update_display();

			/* Send an empty line every 3 seconds to make sure the server still exists */
			if (w++ >= 60) {
				w = 0;
				if (send_nop() < 0) {
					break; /* Out of while loop */
				}
			}
			usleep(50000);
		}
	}

	if (!Quit)
		report(RPT_WARNING, "Server disconnected %d", num_bytes);
	return 0;
}

